src/Features/LanguageServer/Protocol/Handler/InlayHint/InlayHintHandler.cs

```diff
@@ -0,0 +1,112 @@
+            _optionsService = optionsService;
+        }
+
+        public bool MutatesSolutionState => false;
+
+        public bool RequiresLSPSolution => true;
+
+        public TextDocumentIdentifier GetTextDocumentIdentifier(InlayHintParams request)
+            => request.TextDocument;
+
+        public async Task<LSP.InlayHint[]?> HandleRequestAsync(InlayHintParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            var document = context.GetRequiredDocument();
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(request.Range, text);
+
+            var inlineHintService = document.GetRequiredLanguageService<IInlineHintsService>();
+            var options = _optionsService.GetInlineHintsOptions(document.Project.Language);
+
+            var hints = await inlineHintService.GetInlineHintsAsync(document, textSpan, options, displayAllOverride: false, cancellationToken).ConfigureAwait(false);
+            if (hints.IsEmpty)
+            {
+                return Array.Empty<LSP.InlayHint>();
+            }
+
+            using var _ = ArrayBuilder<LSP.InlayHint>.GetInstance(out var inlayHints);
+            var syntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
+            var inlayHintCache = context.GetRequiredLspService<InlayHintCache>();
+
+            // Store the members in the resolve cache so that when we get a resolve request for a particular
+            // member we can re-use the inline hint.
+            var resultId = inlayHintCache.UpdateCache(new InlayHintCache.InlayHintCacheEntry(hints, request.TextDocument, syntaxVersion));
+
+            for (var i = 0; i < hints.Length; i++)
+            {
+                var hint = hints[i];
+                var linePosition = text.Lines.GetLinePosition(hint.Span.Start);
+                var kind = hint.Ranking == 0.0
+                    ? InlayHintKind.Parameter
+                    : InlayHintKind.Type;
+
+                // TextChange is calculated at the same time as the InlineHint,
+                // so it should not need to be resolved.
+                TextEdit[]? textEdits = null;
+                if (hint.ReplacementTextChange.HasValue)
+                {
+                    var textEdit = ProtocolConversions.TextChangeToTextEdit(hint.ReplacementTextChange.Value, text);
+                    textEdits = new TextEdit[] { textEdit };
+                }
+
+                var inlayHint = new LSP.InlayHint
+                {
+                    Position = ProtocolConversions.LinePositionToPosition(linePosition),
+                    Label = ConvertTaggedTextToString(hint.DisplayParts),
+                    Kind = kind,
+                    TextEdits = textEdits,
+                    ToolTip = null,
+                    PaddingLeft = false,
+                    PaddingRight = false,
```

Comment:
so it has padding concepts?  can we use that?  if we see the space on the left/right, we remove it and then set these to trie?